llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2553,6 +2553,42 @@ void CombinerHelper::applyCombineTruncOfExt(
   MI.eraseFromParent();
 }
 
+bool CombinerHelper::matchCombineShlOfAnd(MachineInstr &MI, Register &Reg) {
+  // We're trying to match the following pattern:
+  //   %t = G_AND %x, imm1
+  //   %root = G_SHL %t, imm2
+  // -->
+  //   %root = G_SHL %x, imm2
+  // Where (~imm1 << imm2) = 0
+  assert(MI.getOpcode() == TargetOpcode::G_SHL && "Expected a G_SHL");
+  const Register DstReg = MI.getOperand(0).getReg();
+  const Register SrcReg = MI.getOperand(1).getReg();
+  const LLT SrcTy = MRI.getType(SrcReg);
+  const unsigned Size = SrcTy.getSizeInBits();
+
+  // Try to match shl (and x, imm1), imm2
+  int64_t ShiftImm, AndImm;
+  int64_t Mask = (Size < 64) ? (1LL << Size) - 1 : -1;
```

Comment:
nit: move this mask to just before the use.

Reply:
I also suggest to use unsigned masks whenever is possible. I this way, can we use something like `uint64_t Mask = ~0ULL >> (64-Size);` here?